include/ofi_hook.h

```diff
@@ -67,18 +70,67 @@ extern struct fi_ops hook_fid_ops;
 struct fid *hook_to_hfid(const struct fid *fid);
 struct fid_wait *hook_to_hwait(const struct fid_wait *wait);
 
+/*
+ * TODO
+ * comment from GitHub PR #5052:
+ * "another option would be to store the ini/fini calls in a separate structure
+ * that we reference from struct fi_prov_context. We could even extend the
+ * definition of fi_prov_context with a union that is accessed based on the
+ * prov_type. That might work better if we want to support external hooks,
+ * without the external hook provider needing to implement everything"
+ */
+struct hook_prov_ctx {
+	struct fi_provider	prov;
+	/*
+	 * Hooking providers can override ini/fini calls of a specific fid class
+	 * to override any initializations that the common code may have done.
+	 * For example, this allows overriding any specific op and not having to
+	 * hook into every resource creation call until the point where the op
+	 * can be overridden. Refer to hook_perf for an example.
+	 *
+	 * Note: if a hooking provider overrides any of the resource creation calls
+	 * (e.g. fi_endpoint) directly, then these ini/fini calls won't be
+	 * invoked. */
+	int 			(*ini_fid[HOOK_FI_CLASS_MAX])(struct fid *fid);
+	int 			(*fini_fid[HOOK_FI_CLASS_MAX])(struct fid *fid);
```

Comment:
Do these look okay?

Reply:
Yep -- I think we can do more work on this, but other changes can come later as needed.  My only concern was exposing the change in the public header file.  The atomic definitions had a 'LAST' enum value as well, and those ended up being an issue when adding the collective support.  So, now the 'LAST' values are no longer the last values...